### ISSUE TITLE
Could `image.clean_img` reduce the memory load by using a mask?

### ISSUE THREAD

**Issue by miykael** [nilearn/nilearn#1839]
I have a rather big functional image of dimensions 90 x 120 x 85 x 1055, with `float64` datatype. About 70% of this image is background (i.e. `zeros`). When I want to run `image.clean_img` on this image, my memory load goes up to 15GB.

This is due to the usage of `get_data()`, [here](https://github.com/nilearn/nilearn/blob/master/nilearn/image/image.py#L906) and that we feed all voxels into `signal.clean`. Wouldn't it make sense to introduce additional mask and unmasking steps to compute only on nonempty voxels?

What would the quickest/best way be to create such a mask on the fly? Or should it just be possible to provide a mask to restrict the cleaning?

PS: Such an approach might help other functions as well.

**Comment by bthirion** [comment]
+1 on my side.

**Comment by GaelVaroquaux** [comment]
There are two options possible:

* explicit masking, given as an optional argument
* implicit masking that needs to be guessed.

I think that we should do one, or the other, but not both.

For the second option, I think that we need to have a very simple and non
ambitious strategy, for fear of introducing "magic", ie, behavior that is
not well understood by the user. I think that strategy could be an
argument that list values to be ignored, by default [0, np.nan]. That
should hopefully cover 95% of the use cases.

**Comment by bthirion** [comment]
I think I prefer explicit masking.

**Comment by miykael** [comment]
I think I also prefer explicit masking. I realized that most functional images coming directly from the scanner always have some background noise, therefore `[0, np.nan]` will catch almost nothing in this case.

And as it is just to reduce memory load, an explicit mask could point users to this option, but this is not essential for low-memory-load datasets (and therefore not required).

From the top of your head, are there any other functions where I could implement explicit masking as well?

**Comment by GaelVaroquaux** [comment]
> From the top of your head, are there any other functions where I could
> implement explicit masking as well?

I don't know. Maybe math_img. The reason that I don't know is that I
typically do these operations after after a NiftiMasker: when chaining
operations it is much more efficient to mask only once, and unmask
after.

**Comment by miykael** [comment]
I've created now a new PR for `clean_img`. And yes, you're approach makes sense in a normal environment. I'm currently working in a Nipype environment where everything is modular, so I try to profit from additional input parameters.
